### ISSUE TITLE
Possible system hang during high load

### ISSUE THREAD

**Issue by Pilleo** [hamadmarri/cacule-cpu-scheduler#20]
Hello! Not sure if it was problem of cacule. Using 5.10.1 xanmod with cacule.
So I had running Itellij Idea, Java project, Firefox with hundreds of tabs and a youtube playing. All ram and swap were occupied, but it was all great until it wasn't. Suddenly everything just freezed. earlyoom did not not work, even magic button did not responce. 
Can i do smth to figure out was it cacule problem or not? 
Thank you.

**Comment by hamadmarri** [comment]
Hi @Pilleo 

Could you please confirm that this issue doesn't happen with mainline kernel or with xanmod without cacule?

If this problem is only happening with cacule, please try this fix patch on top of 5.10.1 xanmod/cacule source.
[patch2.zip](https://github.com/hamadmarri/cacule-cpu-scheduler/files/5741853/patch2.zip)

Please let me know if it fixes the issue.

Thanks

**Comment by hamadmarri** [comment]
Another thing that please double check your swap size. Your RAM might got filled with no enough swap area.
Also check the vm.swapness <--- if you haven't change it then don't worry about it (the default value is ok).

**Comment by owl4ce** [comment]
**Sorry, let me leave the topic.**
I want to ask, when will the latest version of CacULE or Cachy for kernel v5.10 be released?

**Comment by hamadmarri** [comment]
Hi @owl4ce 

The patch of 5.9 is fortunately working on 5.10

I am trying to fix the freezing problem and release a version with fixes. Could please try the patch in https://github.com/hamadmarri/cacule-cpu-scheduler/issues/20#issuecomment-751235392
and let me know if it smoother and not mini-freezes under heavy load?

Thank you

**Comment by Pilleo** [comment]
I cannot confirm or refute that it was only with cacule. I have 16 gb ram and 1 gb swap with swappiness 1. And if I am developing Java, than I am always short on memory. Earlier I had frequent freezes, so I installed EarlyOOM. It worked just fine so far. Maybe this time it could not work for some reason, maybe it is smth else. But I did not have that particular problem without cacule with EarlyOOM present. 
[log.txt](https://github.com/hamadmarri/cacule-cpu-scheduler/files/5742535/log.txt)
Hope it will help.

**Comment by raykzhao** [comment]
Hi @Pilleo 

Isn't there a kernel bug at the end of the log? It seems to have something to do with SLUB/usercopy.

**Comment by Pilleo** [comment]
@hamadmarri  
Does it mean it is not a cacule problem? Should we report somewhere else?

**Comment by hamadmarri** [comment]
> @hamadmarri , @raykzhao
> Does it mean it is not a cacule problem? Should we report somewhere else?

Hi @Pilleo 

I don't think it's a cacule issue. I think it is related to the swap since SLUB is a memory allocator. It could be SLUB is not able to allocate more memory. I am sure that cacule has nothing to do with slub or memory issues. 

Thank you

**Comment by Pilleo** [comment]
Is it possible that the problem is n kernel 5.10? In my understanding it should have at least react at magic key, but it was a complete crash

**Comment by hamadmarri** [comment]
> Is it possible that the problem is n kernel 5.10? In my understanding it should have at least react at magic key, but it was a complete crash

Yes, it could be.

**Comment by Salekin-1169** [comment]
I can confirm this, happens during high CPU intensive tasks (xanmod-cacule 5.10 kernel, i3 2 core, 2 thread). 
Example, shader precompile of steam games hangs my system (I saw 4 process `fossilize_replay`, but all of them were using more the 25% of the cpu)

This doesn't occur for memory intensive tasks for me.

**Comment by owl4ce** [comment]
I also experienced the same thing.

**Comment by owl4ce** [comment]
> I can confirm this, happens during high CPU intensive tasks (xanmod-cacule 5.10 kernel, i3 2 core, 2 thread).
> Example, shader precompile of steam games hangs my system (I saw 4 process `fossilize_replay`, but all of them were using more than 25% of the cpu each, thus pushing the system to hang)
> 
> This doesn't occur for memory intensive tasks for me. Also, this wasn't an issue when it was still `Cachy` :)

I also experienced the same thing.

**Comment by raykzhao** [comment]
Hi @hamadmarri @Salekin-1169 @owl4ce 

I just checked the latest xanmod-cacule setting. It doesn't include the starvation fix, and `SCHED_AUTOGROUP` (and therefore `FAIR_GROUP_SCHED`) are enabled. Also it includes the following scheduler tweaks and uses the non-standard 500Hz timer (no such an option in mainline):

```
sysctl_sched_nr_migrate = 256
sysctl_sched_rt_runtime = 980000
```

I just opened an issue at xanmod/linux#112 and mentioned the starvation fix and disabling `FAIR_GROUP_SCHED`.

**Comment by hamadmarri** [comment]
> I can confirm this, happens during high CPU intensive tasks (xanmod-cacule 5.10 kernel, i3 2 core, 2 thread).
> Example, shader precompile of steam games hangs my system (I saw 4 process `fossilize_replay`, but all of them were using more than 25% of the cpu each, thus pushing the system to hang)
> 
> This doesn't occur for memory intensive tasks for me. Also, this wasn't an issue when it was still `Cachy` :)

Hi @Salekin-1169 

This was fixed (hopefully) with the last commit. If you compile from source can you try to patch it with this patch https://github.com/hamadmarri/cacule-cpu-scheduler/issues/15#issuecomment-751220411

Thank you

**Comment by hamadmarri** [comment]
> > I can confirm this, happens during high CPU intensive tasks (xanmod-cacule 5.10 kernel, i3 2 core, 2 thread).
> > Example, shader precompile of steam games hangs my system (I saw 4 process `fossilize_replay`, but all of them were using more than 25% of the cpu each, thus pushing the system to hang)
> > This doesn't occur for memory intensive tasks for me. Also, this wasn't an issue when it was still `Cachy` :)
> 
> I also experienced same thing.

Hi @owl4ce 

Could you please double check that this patch https://github.com/hamadmarri/cacule-cpu-scheduler/issues/15#issuecomment-751220411 is applied?

You can confirm by 
```
cat kernel/sched/fair.c | grep -B2 -A2 "p->se.exec_start = 0;"
```
Output
```
#if !defined(CONFIG_CACULE_SCHED)
        /* We have migrated, no longer consider this task hot */
        p->se.exec_start = 0;
#endif

```


Thank you

**Comment by owl4ce** [comment]
> > > I can confirm this, happens during high CPU intensive tasks (xanmod-cacule 5.10 kernel, i3 2 core, 2 thread).
> > > Example, shader precompile of steam games hangs my system (I saw 4 process `fossilize_replay`, but all of them were using more than 25% of the cpu each, thus pushing the system to hang)
> > > This doesn't occur for memory intensive tasks for me. Also, this wasn't an issue when it was still `Cachy` :)
> > 
> > 
> > I also experienced same thing.
> 
> Hi @owl4ce
> 
> Could you please double check that this patch [#15 (comment)](https://github.com/hamadmarri/cacule-cpu-scheduler/issues/15#issuecomment-751220411) is applied?
> 
> You can confirm by
> 
> ```
> cat kernel/sched/fair.c | grep -B2 -A2 "p->se.exec_start = 0;"
> ```
> 
> Output
> 
> ```
> #if !defined(CONFIG_CACULE_SCHED)
>         /* We have migrated, no longer consider this task hot */
>         p->se.exec_start = 0;
> #endif
> ```
> 
> Thank you

Yes, I patched it.
![2020-12-28-200229_609x351_scrot](https://user-images.githubusercontent.com/53987136/103216069-ae420500-4947-11eb-85c2-476a6aeeda7d.png)

**Comment by hamadmarri** [comment]
I believe there is an issue in v5.10 please see https://github.com/xanmod/linux/issues/111

It could be a bug in mainline kernel v5.10 since the guy how posted the issue in xanmod is not using cacule. I don't think the issue is in xanmod since there is not a big change on xanmod patches from v5.9 to v5.10

Can anyone confirm the freezes is also in v5.9 using the latest cacule patch?

Thank you

**Comment by hamadmarri** [comment]
> > > > I can confirm this, happens during high CPU intensive tasks (xanmod-cacule 5.10 kernel, i3 2 core, 2 thread).
> > > > Example, shader precompile of steam games hangs my system (I saw 4 process `fossilize_replay`, but all of them were using more than 25% of the cpu each, thus pushing the system to hang)
> > > > This doesn't occur for memory intensive tasks for me. Also, this wasn't an issue when it was still `Cachy` :)
> > > 
> > > 
> > > I also experienced same thing.
> > 
> > 
> > Hi @owl4ce
> > Could you please double check that this patch [#15 (comment)](https://github.com/hamadmarri/cacule-cpu-scheduler/issues/15#issuecomment-751220411) is applied?
> > You can confirm by
> > ```
> > cat kernel/sched/fair.c | grep -B2 -A2 "p->se.exec_start = 0;"
> > ```
> > 
> > 
> > Output
> > ```
> > #if !defined(CONFIG_CACULE_SCHED)
> >         /* We have migrated, no longer consider this task hot */
> >         p->se.exec_start = 0;
> > #endif
> > ```
> > 
> > 
> > Thank you
> 
> Yes, I patched it. I am not sure about the problem, when the cpu usage is high and I play the song also sometimes it comes back a few seconds then comes back again. Feels very heavy when all cores are used at 100% usage.
> ![2020-12-28-200229_609x351_scrot](https://user-images.githubusercontent.com/53987136/103216069-ae420500-4947-11eb-85c2-476a6aeeda7d.png)

Can you please try the latest Cacule patch on mainline kernel v5.9 (without xanmod).
Sorry about asking too much compiling, the problem could be not related to cacule.

Thank you

**Comment by hamadmarri** [comment]
This solution is suggested by Alexandre Frade
Thanks to him

>Hamad When executing the nvidia-dkms and mkinitramfs triggers, started the freezes reported by users, even with the fix "remove start_exec = 0", the system normalized without the autogroup: 
`echo 0 |sudo tee /proc/sys/kernel/sched_autogroup_enabled`
all users with this problem, try this solution

**Comment by owl4ce** [comment]
**Hi everyone.**
I can confirm that as said on [this](#issuecomment-751665369) can end this issue.

By disabling
- `SCHED_AUTOGROUP`
- `FAIR_GROUP_SCHED`

**Comment by owl4ce** [comment]
**Hi everyone.**
I can confirm that as said on [this](#issuecomment-751665369) can end this issue.

By disabling this in kernel configuration
- `SCHED_AUTOGROUP`
- `FAIR_GROUP_SCHED`

_Yes it worked for me. Everything is now smooth._

**Comment by Salekin-1169** [comment]
@hamadmarri @raykzhao @owl4ce sorry for the late reply. I updated the latest xanmod-cacule package and tested.
Running steam shader precompile with youtube running in background.
my system didn't come to a complete hang like before, but it was still unusable (audio skips in youtube, screen lags)

After that, I changed the value of `kernel.sched_interactivity_factor' to `50` and the system immediately became responsive.

**Comment by Salekin-1169** [comment]
I can even record what's happening currently on my pc now, which is impossible with the default value of `kernel.sched_interactivity_factor`

![Uploading Peek 2020-12-29 22-13.gif…]()

**Comment by hamadmarri** [comment]
> @hamadmarri @raykzhao @owl4ce sorry for the late reply. I updated the latest xanmod-cacule package and tested.
> Running steam shader precompile with youtube running in background.
> my system didn't come to a complete hang like before, but it was still unusable (audio skips in youtube, screen lags)
> 
> After that, I changed the value of `kernel.sched_interactivity_factor` to `50` and the system immediately became responsive.

What was the default value? 32768?

**Comment by Salekin-1169** [comment]
@hamadmarri 32768

**Comment by Salekin-1169** [comment]
@hamadmarri I have a comparatively weak cpu (i3 2 physical + 2 logical cores), so I think the default `32768` isn't suitable for my pc, while other people didn't face issue with this.

**Comment by hamadmarri** [comment]
> @hamadmarri I have a comparatively weak cpu (i3 2 physical + 2 logical cores), so I think the default `32768` isn't suitable for my pc, while other people didn't face issue with this.


Hi @Salekin-1169 

https://github.com/hamadmarri/cacule-cpu-scheduler/blob/06a5ed115674bf63c31ab5ba1acf5f98cc21338e/patches/CacULE/v5.9/cacule5.9.patch#L382

I believe the problem is in this line. It seems that a task that has high run time will get lower value! I am not sure if original ULE used it this way. I will try one more fix. Also I will add reset_life_time as same as cachy has.

 The task flips to be non interactive where it should have a score somehow closer to it's previous score, but from the math I can see that the score jumps to the lowest and gradually starts to gain some score.

**Comment by hamadmarri** [comment]
[interactivity_score_fix.zip](https://github.com/hamadmarri/cacule-cpu-scheduler/files/5751880/interactivity_score_fix.zip)

Can you please try this patch. I fixed the interactivity score equation. It is now very similar to Cachy/HRRN

**Comment by raykzhao** [comment]
Hi @hamadmarri 

I think the `cacule_max_lifetime` in the new patch should be better tunable via sysctl. Also it doesn't seem to include the starvation fix.

**Comment by Salekin-1169** [comment]
> [interactivity_score_fix.zip](https://github.com/hamadmarri/cacule-cpu-scheduler/files/5751880/interactivity_score_fix.zip)
> 
> Can you please try this patch. I fixed the interactivity score equation. It is now very similar to Cachy/HRRN

Thank you for your quick fix :bow: 

I apologize, because I'm still relatively new to linux and don't know how to compile custom kernel yet. I'll study about it this weekend and check if the latest patch fixes the issue, then let you know :vulcan_salute:

**Comment by raykzhao** [comment]
Hi @Salekin-1169 

I'm not sure which Linux distro you are using. Generally speaking, you may try:

1. Download the latest kernel source from https://www.kernel.org/
2. Copy the patch file to the kernel source folder, and run `patch -p1 -i interactivity_score_fix.patch`
3. Get your current kernel config file. Usually it should be located at `/proc/config.gz`. If not, try to run `modprobe configs` first. 
4. In the kernel source folder, run `zcat /proc/config.gz > .config`
5. Run `make menuconfig` and make sure you disable the following: 
```
General Setup-->Automatic process group scheduling
General Setup-->Control Group support-->CPU controller-->Group scheduling for SCHED_OTHER
```
You may also want to distinguish with the existing kernel by appending some suffix at `General Setup-->Local version - append to kernel release`.
6. Run `make -jx`, where x is the number of CPU cores. For example, if you have a 4-core CPU, you may run `make -j4`.
7. Run `make install` and `make modules_install`.
8. Update your bootloader configuration. Based on your distro, the instructions will be different.
9. If you are using any out-of-tree modules e.g. nvidia proprietary drivers, you also need to recompile or reinstall them. Depend on your distro, the instructions will be different.

For the bootloader and out-of-tree modules, you should lookup the instructions of your specific Linux distro. Some distros may also have a guide on how to build custom kernels.

**Comment by hamadmarri** [comment]
> Hi @hamadmarri
> 
> I think the `cacule_max_lifetime` in the new patch should be better tunable via sysctl. Also it doesn't seem to include the starvation fix.

Hi @raykzhao 

The starvation fix actually is not a good solution since a task can keep migrating even if it is cache hot. I think the fix in interactivity score is better approach. Also I was mistaken about vruntime getting reset, only exec_start got reset on migration, which updated right away in select_next_task func.

Thank you

**Comment by raykzhao** [comment]
Hi @hamadmarri 

Although the new fix seems better than the vanilla CacULE on my machine, unfortunately it doesn't feel better than the original Cachy or the CacULE with either the smoother or the starvation fix. Somehow it's slightly more glitchy on my machine under heavy load. 

I still think both Cachy and CacULE should be kept in a single patch (#19), since it's difficult to find a scheduling policy that is suitable for everyone. On my machine the original Cachy is definitely the winner, but it may not be the case for others.

**Comment by owl4ce** [comment]
Same, on my machine Cachy is more responsive but not under heavy load, CacULE is much more responsive if my machine is running heavy duty under 100% pressure of all cores.

**Comment by hamadmarri** [comment]
> Hi @hamadmarri
> 
> Although the new fix seems better than the vanilla CacULE on my machine, unfortunately it doesn't feel better than the original Cachy or the CacULE with either the smoother or the starvation fix. Somehow it's slightly more glitchy on my machine under heavy load.
> 
> I still think both Cachy and CacULE should be kept in a single patch (#19), since it's difficult to find a scheduling policy that is suitable for everyone. On my machine the original Cachy is definitely the winner, but it may not be the case for others.

Hi @raykzhao 

What I am worried about is that the problem is not on cachy nor on cacule. Maybe it's on v5.9 and v5.10. Could please confirm that cachy has no issues on both v5.9 and v5.10?

Thank tou

**Comment by raykzhao** [comment]
Hi @hamadmarri 

I can confirm that there is no difference on my machine with Cachy scheduler between 5.9 and 5.10 kernels.

**Comment by raykzhao** [comment]
Hi @Salekin-1169 

I just found that the interactivity score fix should already be pushed to xanmod/linux@16e99a8c42f36bc83be0af522e26d16e1cd64e98 and therefore 5.10.4-xanmod1-cacule should include the fix. You don't need to build your own custom kernel now.

**Comment by Salekin-1169** [comment]
@hamadmarri @raykzhao sorry about the delayed reply. I tested the latest fix, and it fixed my issue completely. Also, I tested on a freshly reinstalled system, so all the values are set to default.
![sched_interactivity](https://user-images.githubusercontent.com/49670756/103441882-b5d51700-4c83-11eb-8549-54b311311eba.png)


I only faced some minor audio lags (very minimal) during recording, but other than that, everything was butter smooth.
![Peek 2021-01-01 22-53](https://user-images.githubusercontent.com/49670756/103441941-2b40e780-4c84-11eb-9aa6-60b3288007cb.gif)

Thank you so much for your support, the issue is completely fixed for me :bow:
Wish you an Happy New Year :beers:

**Comment by hamadmarri** [comment]
> @hamadmarri @raykzhao sorry about the delayed reply. I tested the latest patch, and it fixed my issue completely. Also, I tested on a freshly reinstalled system, so all the values are set to default.
> ![sched_interactivity](https://user-images.githubusercontent.com/49670756/103441882-b5d51700-4c83-11eb-8549-54b311311eba.png)
> 
> I only faced some minor audio lags (very minimal) during recording, but other than that, everything was butter smooth.
> ![Peek 2021-01-01 22-53](https://user-images.githubusercontent.com/49670756/103441941-2b40e780-4c84-11eb-9aa6-60b3288007cb.gif)
> 
> Thank you so much for your support, the issue is completely fixed for me bow
> Wish you all a very Happy New Year beers

Glad to hear it is working good now.
Happy new year @Salekin-1169 .

**Comment by Salekin-1169** [comment]
I learned a lot so much about how ULE works from [this article](https://www.usenix.org/conference/atc18/presentation/bouron#:~:text=We%20compare%20ULE%20and%20CFS,are%20normally%20scheduled%20by%20CFS.&text=The%20more%20complex%20load%20balancing,balance%20in%20the%20long%20run.) and [slide](https://www.usenix.org/sites/default/files/conference/protected-files/atc18_slides_bouron.pdf)

ULE by design will cause starvation, but will yield better throughput it seems? I had completely opposite idea about it.

**Comment by hamadmarri** [comment]
> I learned so much about how ULE works from [this article](https://www.usenix.org/conference/atc18/presentation/bouron#:~:text=We%20compare%20ULE%20and%20CFS,are%20normally%20scheduled%20by%20CFS.&text=The%20more%20complex%20load%20balancing,balance%20in%20the%20long%20run.) and [slide](https://www.usenix.org/sites/default/files/conference/protected-files/atc18_slides_bouron.pdf) (Can't find an appropriate place to share, so posting here temporarily).
> 
> ULE by design will cause starvation, but will yield better throughput it seems? I had completely opposite idea about it.

Hi @Salekin-1169 

This study unfortunately is an implementation of ULE on top of CFS. They replaced some functions in CFS with ULE way of their implementation. I don't call this a fair comparison. Their conclusion based on the stats and results they had from this unfair implementation in which they conclude that ULE could lead to starvation.

Yes, I think CFS is more advance than ULE, but I am not really convinced that ULE lead to starvation based on a single unfair study. I have tried FreeBSD, it is smoother and didn't faced any kind of starvation while stressing the system with many kind of tests. ULE code is cleaner, 10x less than CFS, and probably faster, however, magically, CFS provides slightly higher throughput than ULE.

Thank you

**Comment by Salekin-1169** [comment]
@hamadmarri thank you so much for your explanation. I'm still learning about it. CFS has about a decade of optimizations behind it, so for that it's throughput is higher I guess ?

Just curious, how different is Cacule from the FreeBSD implementation of ULE? Can I read up about it somewhere?

**Comment by hamadmarri** [comment]
> @hamadmarri thank you so much for your explanation. I'm still learning about it. CFS has about a decade of optimizations behind it, so for that it's throughput is higher I guess ?
> 
> Just curious, how different is Cacule from the FreeBSD implementation of ULE? Can I read up about it somewhere?

CFS is a result of previous crazy approaches. One of them was the scheduler made by the genius Con Kolivas (the author of bfs and muqss) - his scheduler then was Staircase scheduler. Ingo Molnár made CFS which is inspired by Kolivas's work. FreeBSD didn't have a crazy guy like Kolivas.

>Just curious, how different is Cacule from the FreeBSD implementation of ULE? Can I read up about it somewhere?

https://github.com/hamadmarri/cacule-cpu-scheduler#the-cacule-interactivity-score

I have only implemented the interactivity (IS) score (see Figure 1 https://github.com/hamadmarri/cacule-cpu-scheduler/blob/master/helper%20docs%20for%20kernel%20dev/FreeBSD/ULE.pdf)

My implementation of IS is not similar to ULE, it is more like the Cachy/HRRN way.
ULE uses 2 runqueues, and based on the IS math, the task is placed to some runqueue. They also use multiple levels queues for priority. Where I just use the CFS's vruntime calculations which affect the task priority on vruntime value. Which therefore affects the total run time of the process when calculating HRRN or IS

I used a shortcut approach to adapt IS to CFS. ULE is totally different than CFS. I remember that ULE is about 3k LOC only, where CFS is absolutely is more than 25k LOC (I counted only 4 files in CFS no all)
```
Here just 4 files LOC in linux I deal with daily

❯ cat kernel/sched/fair.c | wc -l
11872

~/dev/linux/linux rdb
❯ cat kernel/sched/core.c | wc -l
8498

~/dev/linux/linux rdb
❯ cat kernel/sched/sched.h | wc -l
2643

~/dev/linux/linux rdb
❯ cat include/linux/sched.h | wc -l
2075
```

Thank you
